Add implementation for vapp network firewall rules resource (#511)
terraform-providers/terraform-provider-vcd: website/docs/r/vapp_firewall_rules.html.markdown

---
layout: "vcd"
page_title: "vCloudDirector: vcd_vapp_firewall_rules"
sidebar_current: "docs-vcd-resource-vapp-firewall-rules"
description: |-
  Provides a vCloud Director vApp Firewall resource. This can be used to create, modify, and delete firewall settings and rules.
---

# vcd\_vapp\_firewall\_rules

Provides a vCloud Director vApp Firewall resource. This can be used to create,

> modify, and delete firewall settings and rules.

Reviewer: ```suggestion
modify, and delete firewall settings and rules in a vApp network.
```
Also, please make a link out of "vApp network" and point to the documentation page of it, so we link these two pages between each other.

Author: added

> ~> **Note:** Using this resource overrides any existing firewall rules on vApp network.

Reviewer: Useful note! What about making it a warning, so it stands out more?

Author: changed

> !> **Warning:** Using this resource overrides any existing firewall rules on vApp network.

Author: added

> !> **Warning:** Using this resource overrides any existing firewall rules on vApp network. It's recommended to have only one resource per vApp. 

Reviewer: >It's recommended to have only one resource per vApp. 

Maybe Terraform has a way to enforce this?

## Example Usage

```hcl
resource "vcd_vapp" "web" {
  name = "web"
}
​
resource "vcd_vapp_network" "vapp-net" {
  name               = "my-vapp-net"
  vapp_name          = vcd_vapp.web.name
  org_network_name   = "my-vdc-int-net"
  gateway            = "192.168.2.1"
  netmask            = "255.255.255.0"
  dns1               = "192.168.2.1"
​
  static_ip_pool {
    start_address = "192.168.2.51"
    end_address   = "192.168.2.100"
  }

> ​

Reviewer: Extra line.

Author: removed

}

> resource "vcd_vapp_firewall_rules" "vapp_fw" {

Reviewer: Missing empty line above. Terraform fmt on samples missing?

Author: added. Terraform fmt  doesn't correct such things.

  vapp_id        = vcd_vapp.TestAccVcdVAppForInsert.id
  network_id     = vcd_vapp_network.vapp-net.id
  default_action = "drop"

  rule {
    description      = "drop-ftp-out"
    policy           = "drop"
    protocol         = "tcp"
    destination_port = "21"
    destination_ip   = "any"
    source_port      = "any"
    source_ip        = "10.10.0.0/24"
  }

  rule {
    description      = "allow-outbound"
    policy           = "allow"
    protocol         = "any"
    destination_port = "any"
    destination_ip   = "any"
    source_port      = "any"
    source_ip        = "10.10.0.0/24"
  }
}
```

## Argument Reference

The following arguments are supported:

> * `org` - (Optional; *v2.0+*) The name of organization to use, optional if defined at provider level. Useful when connected as sysadmin working across different organisations.

Author: removed

* `vdc` - (Optional; *v2.0+*) The name of VDC to use, optional if defined at provider level.

> * `vapp_id` - (Required) The identifier of vApp.

Reviewer: vApp - please add a link to doc page here too.

Author: added

Author: added

> * `network_id` - (Required) The identifier of vApp network.

Reviewer: vApp network - please add a link to doc page here too.

Author: added

Author: added

* `default_action` - (Required) Either "allow" or "drop". Specifies what to do should none of the rules match
* `log_default_action` - (Optional) Flag to enable logging for default action. Default value is false
* `rule` - (Optional) Configures a firewall rule; see [Rules](#rules) below for details.

<a id="rules"></a>
## Rules

Each firewall rule supports the following attributes:

* `description` - (Optional) Name of the firewall rule

> * `enabled` - (Optional) 'true' value will enable firewall rule

Reviewer: Wrong quotes - these do not render as code.

Author: fixed

> * `description` - (Optional) Name of the firewall rule.

Reviewer 2: Ignore. It was before it was changed.

> * `enabled` - (Optional) `true` value will enable firewall rule.

Author: added

* `policy` - (Optional) Specifies what to do when this rule is matched. Either "allow" or "drop"
* `protocol` - (Optional) The protocol to match. One of "tcp", "udp", "icmp", "any"" or "tcp&udp"
* `destination_port` - (Optional) The destination port to match. Either a port number or "any"
* `destination_ip` - (Optional) The destination IP to match. Either an IP address, IP range or "any"
* `destination_vm_id` - (Optional) Destination VM identifier
* `destination_vm_ip_type` - (Optional) The value can be one of: `assigned` - assigned internal IP be automatically choosen. `NAT`: NATed external IP will be automatically choosen
* `destination_vm_nic_id` - (Optional) VM NIC ID to which this rule applies
* `source_port` - (Optional) The source port to match. Either a port number or "any"
* `source_ip` - (Optional) The source IP to match. Either an IP address, IP range or "any"
* `source_vm_id` - (Optional) Source VM identifier

> * `source_vm_ip_type` - (Optional) The value can be one of: `assigned` - assigned internal IP be automatically choosen. `NAT`: NATed external IP will be automatically choosen

Reviewer: ```suggestion
* `source_vm_ip_type` - (Optional) The value can be one of: `assigned` - use assigned internal IP, `NAT` - use NATed external IP.
```
Also, same ask for `destination_vm_ip_type` above.

Author: updated

Reviewer: Somehow still asks for a cleaner look... :D Maybe @dataclouder or @Didainius can suggest a better syntax for this sentence.

Reviewer: Looking at this again. In the first version you had stated that **IP will be automatically chosen**. How does this actually work? Would it add value if have this statement in the description after all?

Author: Description is from API

Author: ![image](https://user-images.githubusercontent.com/41415459/85097662-e500cb00-b200-11ea-8c81-9f27e6eeeef5.png)

Reviewer: Got it!

https://code.vmware.com/apis/912/vmware-cloud-director/doc/doc/types/VmSelectionType.html

> The value can be one of:- assigned: assigned internal IP be automatically choosen. NAT: NATed external IP will be automatically choosen. 

So, based on this, a new suggestion (at least we make punctuation clearer and fix spelling :) ):

```
* `source_vm_ip_type` - (Optional) The value can be one of: `assigned` - assigned internal IP will be automatically chosen, `NAT` - NATed external IP will be automatically chosen.
```

Author: done

* `source_vm_nic_id` - (Optional) VM NIC ID to which this rule applies

> * `enable_logging`- (Optional) 'true' value will enable rule logging. Default is false

Reviewer: - Please add dots at the end of bullet point sentences above.
- Instead of quotes for constants (e.g. "allow", "tcp", "any") format them as code (i.e. `allow`, `tcp`, `any`).
- Format boolean values (i.e. `false`) as code.

Author: done

## Importing

~> **Note:** The current implementation of Terraform import can only import resources into the state.
It does not generate configuration. [More information.](https://www.terraform.io/docs/import/)

An existing an vApp network firewall rules can be [imported][docs-import] into this resource
via supplying the full dot separated path to vapp network. An example is
below:

```
terraform import vcd_vapp_firewall_rules.my-rules my-org.my-vdc.vapp_name.network_name
```
or using IDs:
```

> terraform import vcd_vapp_firewall_rules.my-rules my-org.my-vdc.vapp_id.network_id

Reviewer: How the end user can get the vApp network ID? If no easy way, what about providing the list function like we had for other resources?
CC: @dataclouder

Reviewer 3: +1 
I missed this in my review.

Author: added list functionality

Reviewer: This is great, thanks!

---
layout: "vcd"
page_title: "vCloudDirector: vcd_vapp_firewall_rules"
sidebar_current: "docs-vcd-resource-vapp-firewall-rules"
description: |-
  Provides a vCloud Director vApp Firewall resource. This can be used to create, modify, and delete firewall settings and rules.
---

# vcd\_vapp\_firewall\_rules

Provides a vCloud Director vApp Firewall resource. This can be used to create,

> modify, and delete firewall settings and rules in a [vApp network](/docs/providers/vcd/r/vapp_network.html).

Reviewer: Could we add a sentence on which vApp network types the FW can be created and on which not? Currently, you have to find out the hard way that the isolated network doesn't support it:
```
vcd_vapp_firewall_rules.vapp_fw: Creating...

Error: error setting firewall rules: provided network isn't connecd to org network or isn't fenced

  on yorktown.tf line 359, in resource "vcd_vapp_firewall_rules" "vapp_fw":
 359: resource "vcd_vapp_firewall_rules" "vapp_fw" {
```

Author: added

Reviewer: Word "to" missing:

> Firewall rules can be applied to networks connected **to** Org network or vApp networks which are fenced.

Author: added

* `source_vm_nic_id` - (Optional) VM NIC ID to which this rule applies.
* `enable_logging`- (Optional) `true` value will enable rule logging. Default is `false`.

## Importing

~> **Note:** The current implementation of Terraform import can only import resources into the state.
It does not generate configuration. [More information.](https://www.terraform.io/docs/import/)

An existing an vApp network firewall rules can be [imported][docs-import] into this resource
via supplying the full dot separated path to vapp network. An example is
below:

```
terraform import vcd_vapp_firewall_rules.my-rules my-org.my-vdc.vapp_name.network_name
```
or using IDs:
```
terraform import vcd_vapp_firewall_rules.my-rules my-org.my-vdc.vapp_id.network_id
```

NOTE: the default separator (.) can be changed using Provider.import_separator or variable VCD_IMPORT_SEPARATOR

[docs-import]:https://www.terraform.io/docs/import/

After that, you can expand the configuration file and either update or delete the vApp network rules as needed. Running `terraform plan`
at this stage will show the difference between the minimal configuration file and the vApp network rules stored properties.

> ### Listing vApp network IDs

Reviewer: ```suggestion
### Listing vApp Network IDs
```

Author: fixed

If you want to list IDs there is a special command **`terraform import vcd_vapp_firewall_rules.imported list@org-name.vcd-name.vapp-name`**

> where `org-name` is the organization used, `vdc-name` is vDC name and `vapp-name` is vAPP name. 

Reviewer: ```suggestion
where `org-name` is the organization used, `vdc-name` is VDC name and `vapp-name` is vApp name. 
```

Author: fixed

The output for this command should look similar to the one below:

```shell
$ terraform import vcd_vapp_firewall_rules.imported list@org-name.vdc-name.vapp-name

> vcd_vm_internal_disk.imported: Importing from ID "list@org-name.vdc-name.vapp-name"...

Reviewer: vcd_vm_internal_disk?

Author: fixed

Retrieving all vApp networks by name
No	vApp ID                                                 ID                                      Name	
--	-------                                                 --                                      ----	
1	urn:vcloud:vapp:77755b9c-5ec9-41f7-aceb-4cf158786482	0027c6ae-7d59-457e-b33e-a89e97f0bdc1	Net2
2	urn:vcloud:vapp:77755b9c-5ec9-41f7-aceb-4cf158786482	36986073-8051-4f6d-a1c6-bda648bdf6ba	Net1      		

Error: resource was not imported! resource id must be specified in one of these formats:
'org-name.vdc-name.my-independent-disk-id' to import by rule id

> 'list@org-name.vdc-name.my-independent-disk-name' to get a list of disks with their IDs

Reviewer: independent-disk? There are multiple places in the output of list command where left-over disk copy/pastes are. Please re-review whole output of the list command:

```
lvirbalas$ terraform import vcd_vapp_firewall_rules.vapp_fw list@demo-org.demo-vdc.demo-web
vcd_vapp_firewall_rules.vapp_fw: Importing from ID "list@demo-org.demo-vdc.demo-web"...
Retrieving all vApp networks by name
No	vApp ID							ID					Name
--	-------							--					----
1	urn:vcloud:vapp:2521e464-beec-4226-9fcf-3b37b240076c	07c10396-cd28-4c0b-a72a-c4d4f4697697	demo-net-web
2	urn:vcloud:vapp:2521e464-beec-4226-9fcf-3b37b240076c	10855c07-6685-4635-81ff-c5bb8851a852	demo-vapp-net-w-org
3	urn:vcloud:vapp:2521e464-beec-4226-9fcf-3b37b240076c	e836f48e-0172-49b3-88b6-f57e5d567567	demo-vapp-net-isolated

Error: resource was not imported! resource id must be specified in one of these formats:
'org-name.vdc-name.my-independent-disk-id' to import by rule id
'list@org-name.vdc-name.my-independent-disk-name' to get a list of disks with their IDs
```

Author: fixed

Reviewer: Here:
<img width="876" alt="Screenshot 2020-06-18 at 19 25 49" src="https://user-images.githubusercontent.com/11330689/85047349-58232680-b19a-11ea-848f-92c58f2de427.png">
Second example should be formatted same way as the first one.

Author: ![image](https://user-images.githubusercontent.com/41415459/85097157-50499d80-b1ff-11ea-940f-3f56f50c5a43.png)
done
